onnxruntime/core/providers/cuda/tensor/scatter_elements_impl.cu

```diff
@@ -186,49 +206,109 @@ Status ScatterElementsImpl(
+          func);
     }
 
     int blocksPerGrid = gsl::narrow_cast<int>(CeilDiv(indices_size, GridDim::maxThreadsPerBlock));
     _ScatterElementsKernel<T, Tin><<<blocksPerGrid, GridDim::maxThreadsPerBlock, 0>>>(
         rank, input_data, buffer_input_dims, buffer_input_strides,
         indices_data, indices_size, buffer_indices_dims, fdm_indices_strides,
-        updates, axis, output_data);
+        updates, axis, output_data, func);
   }
   return Status::OK();
 }
 
-#define SPECIALIZED_TINDEX_IMPL(T, TIndex)                        \
-  template Status ScatterElementsImpl<T, TIndex>(                 \
-      const int rank,                                             \
-      const T* input_data,                                        \
-      const int64_t input_size,                                   \
-      TArray<int64_t>& buffer_input_dims,    \
-      TArray<int64_t>& buffer_input_strides, \
-      const TIndex* indices_data,                                 \
-      const int64_t indices_size,                                 \
-      TArray<int64_t>& buffer_indices_dims,  \
-      TArray<fast_divmod>& indices_strides,  \
-      const T* updates,                                           \
-      const int axis,                                             \
+template <typename T, typename Tin>
+Status ScatterElementsImpl(
+    const int rank,
+    const T* input_data,
+    const int64_t input_size,
+    TArray<int64_t>& buffer_input_dims,
+    TArray<int64_t>& buffer_input_strides,
+    const Tin* indices_data,
+    const int64_t indices_size,
+    TArray<int64_t>& buffer_indices_dims,
+    TArray<fast_divmod>& fdm_indices_strides,
+    const T* updates,
+    const int axis,
+    T* output_data) {
+  return ScatterElementsImplInternal(rank, input_data, input_size, buffer_input_dims,
+                                     buffer_input_strides, indices_data, indices_size, buffer_indices_dims, fdm_indices_strides,
+                                     updates, axis, output_data, Func_Assignment<T>());
+}
+
+template <typename T, typename Tin>
+Status GatherElementsGradImpl(
+    const int rank,
+    TArray<int64_t>& buffer_input_dims,
+    TArray<int64_t>& buffer_input_strides,
+    const Tin* indices_data,
+    const int64_t indices_size,
+    TArray<int64_t>& buffer_indices_dims,
+    TArray<fast_divmod>& fdm_indices_strides,
+    const T* updates,
+    const int axis,
+    T* output_data) {
+  // Give output_data as the input_data parameter by intention,
+  // to skip input_data copy, which is not applicable for GatherElementsGrad.
+  return ScatterElementsImplInternal(rank, output_data, 0,
+                                     buffer_input_dims, buffer_input_strides, indices_data,
+                                     indices_size, buffer_indices_dims, fdm_indices_strides,
+                                     updates, axis, output_data, Func_AtomicAdd<T>());
+}
```

Comment:
This means that we are always building GatherElementsGradImpl in the inference binary 

Anyway to refactor them and make GatherElementsGradImpl in a separated file in the orttraining folder?

Reply:
Put GatherElementsGradImpl in a separated file, and use ENABLE_TRAINING to control the build sutff in scatter_elements_impl.cu